cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,489 @@
+ * first available spot that's large enough.  Puts the reserved block at the end
+ * of the available one and, if necessary, removes the available one from the
+ * list of free blocks.
+ */
+void *MemoryManager::MallocArena::Malloc(const uint32_t size) {
+  assert(size > 0);
+
+  // Control word first, block type tag last
+  int32_t total_size = sizeof(ReservedBlockCtl) + size + 1;
+  if (total_size < kMinBlockSize)
+    total_size = kMinBlockSize;
+
+  bool wrapped = false;
+  // Generally: p = LINK(q)
+  AvailBlockCtl *q = rover_;
+  AvailBlockCtl *p;
+  do {
+    p = q->GetNextPtr(arena_);
+    if (p->size >= total_size)
+      break;
+    if (p == head_avail_) {
+      if (wrapped)
+        return NULL;
+      wrapped = true;
+    }
+    q = p;
+  } while (true);
+
+  // As of here: we found a sufficiently large free block at p following q
+  no_reserved_++;
+
+  int32_t remaining_size = p->size - total_size;
+  // Avoid creation of very small blocks
+  if (remaining_size < kMinBlockSize) {
+    total_size += remaining_size;
+    remaining_size = 0;
+  }
+
+  // Update the list of available blocks
+  if (remaining_size == 0) {
+    // Remove free block p from the list of available blocks
+    q->link_next = p->link_next;
+    q->GetNextPtr(arena_)->link_prev = q->ConvertToLink(arena_);
+  } else {
+    p->size = remaining_size;
+    void *upper_tag =
+      reinterpret_cast<char *>(p) + remaining_size - sizeof(AvailBlockTag);
+    new (upper_tag) AvailBlockTag(remaining_size);
+  }
+  rover_ = q->GetNextPtr(arena_);
+
+  // Place the new allocation, which also sets the block type tag at the end
+  char *new_block = reinterpret_cast<char *>(p) + remaining_size;
+  new (new_block) ReservedBlockCtl(total_size);
+  return new_block + sizeof(ReservedBlockCtl);
+}
+
+
+/**
+ * The arena starts with a pointer to this followed by the AvailBlockCtl of
+ * head_avail_, followed by a reserved tag to prevent it from being merged,
+ * followed by a free block spanning the arena until the end tag.  The end tag
+ * is a single negative int, which mimics another reserved block.
+ */
+MemoryManager::MallocArena::MallocArena()
+  : arena_(reinterpret_cast<char *>(sxmmap_align(kArenaSize)))
+  , head_avail_(reinterpret_cast<AvailBlockCtl *>(arena_ + sizeof(void *)))
+  , rover_(head_avail_)
+  , no_reserved_(0)
+{
+  int32_t usable_size = kArenaSize -
+    (sizeof(void *) + sizeof(AvailBlockCtl) + 1 + sizeof(int32_t));
+  *reinterpret_cast<MallocArena **>(arena_) = this;
+  AvailBlockCtl *free_block =
+    new (arena_ + sizeof(void *) + sizeof(AvailBlockCtl) + 1) AvailBlockCtl();
+  free_block->size = usable_size;
+  free_block->link_next = free_block->link_prev =
+    head_avail_->ConvertToLink(arena_);
+
+  head_avail_->size = 0;
+  head_avail_->link_next = head_avail_->link_prev =
+    free_block->ConvertToLink(arena_);
+
+  // Prevent succeeding blocks from merging
+  *(reinterpret_cast<char *>(free_block) - 1) = kTagReserved;
```

Comment:
Ah! There are actually two `AvailBlockCtl`s! Only that the constructor of the one at `head_avail_` is never called! Furthermore `head_avail_` is only a sentinel that serves as a list-head, right?